server/src/main/java/org/elasticsearch/common/logging/ChunkedLoggingStream.java

```diff
@@ -0,0 +1,136 @@
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.common.logging;
+
+import org.apache.logging.log4j.Level;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.common.unit.ByteSizeUnit;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.charset.StandardCharsets;
+import java.util.Base64;
+import java.util.zip.GZIPOutputStream;
+
+/**
+ * An {@link OutputStream} which Gzip-compresses the written data, Base64-encodes it, and writes it in fixed-size chunks to a logger. This
+ * is useful for debugging information that may be too large for a single log message and/or which may include data which cannot be
+ * recorded faithfully in plain-text (e.g. binary data or data with significant whitespace).
+ */
+public class ChunkedLoggingStream extends OutputStream {
+
+    static final int CHUNK_SIZE = ByteSizeUnit.KB.toIntBytes(2);
+
+    /**
+     * Create an {@link OutputStream} which Gzip-compresses the written data, Base64-encodes it, and writes it in fixed-size (2kiB) chunks
+     * to the given logger. If the data fits into a single chunk then the output looks like this:
+     *
+     * <pre>
+     * $PREFIX (gzip compressed and base64-encoded): H4sIAAAAA...
+     * </pre>
+     *
+     * If there are multiple chunks then they are written like this:
+     *
+     * <pre>
+     * $PREFIX [part 1]: H4sIAAAAA...
+     * $PREFIX [part 2]: r38c4MBHO...
+     * $PREFIX [part 3]: ECyRFONaL...
+     * $PREFIX [part 4]: kTgm+Qswm...
+     * $PREFIX (gzip compressed, base64-encoded, and split into 4 parts on preceding log lines)
+     * </pre>
+     *
+     * @param logger The logger to receive the chunks of data.
+     * @param level  The log level to use for the logging.
+     * @param prefix A prefix for each chunk, which should be reasonably unique to allow for reconstruction of the original message even if
+     *               multiple such streams are used concurrently.
+     */
+    public static OutputStream create(Logger logger, Level level, String prefix) throws IOException {
+        return new GZIPOutputStream(Base64.getEncoder().wrap(new ChunkedLoggingStream(logger, level, prefix)));
+    }
+
+    private final Logger logger;
+    private final Level level;
+    private final String prefix;
+
+    private int chunk;
+    private int offset;
+    private boolean closed;
+    private final byte[] buffer = new byte[CHUNK_SIZE];
+
+    ChunkedLoggingStream(Logger logger, Level level, String prefix) {
+        this.logger = logger;
+        this.level = level;
+        this.prefix = prefix;
+    }
+
+    private void flushBuffer() {
+        assert closed || offset == CHUNK_SIZE : offset;
+        assert offset >= 0 && offset <= CHUNK_SIZE : offset;
+        chunk += 1;
+
+        final var chunkString = new String(buffer, 0, offset, StandardCharsets.ISO_8859_1);
```

Comment:
I guess the standard charset here encompasses all the safe bytes mentioned below?

Reply:
Yes, [ISO-8859-1](https://en.wikipedia.org/wiki/ISO/IEC_8859-1) agrees with US-ASCII (and hence UTF-8) on these bytes, and is the encoding that Java uses internally which makes this the most efficient way to construct a `String` from some bytes.